File: DataStructure/Queue/133_CloneGraph.py
# 克隆图
'''
# Node 的定义
class Node(object):
    def __init__(self, val = 0, neighbors = []):
        self.val = val
        self.neighbors = neighbors
'''
class Solution(object):

    def __init__(self):
        self.visited = {}
    
    def cloneGraph(self, node):
        if not node:
            return node # node是一个节点 如果不是一个节点
        if node in self.visited: # 这个节点访问过
            return self.visited[node] # ？？
        clone_node = Node(node.val, []) # 克隆一个node
        self.visited[node] = clone_node 

        if node.neighbors:
            clone_node.neighbors = [self.cloneGraph(n) for n in node.neighbors]

        return clone_node

File: DataStructure/Queue/133_CloneGraph.java
package LeetCode.Graphs;

import java.util.List;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.HashMap;
import java.util.Queue;

import javax.swing.plaf.nimbus.NimbusLookAndFeel;

class Node {
    public int val;
    public List<Node> neighbors;

    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }

    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }

    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}

// BFS 外加一个 hashtable 
class Solution {
    public Node cloneGraph(Node node) {
        if (node == null)
            return null;
        Queue<Node> q = new LinkedList<>();
        HashMap<Node, Node> map = new HashMap<>();
        q.offer(node);
        map.put(node, new Node(node.val));

        while (!q.isEmpty()) {
            Node cur = q.poll();

            for (Node neighbor : cur.neighbors) {
                if (!map.containsKey(neighbor)) {
                    map.put(neighbor, new Node(neighbor.val));
                    q.offer(neighbor);
                }
                map.get(cur).neighbors.add(map.get(neighbor));
            }
        }
        return map.get(node);
    }
}
// 通过了